lib/web_ui/lib/src/engine/text_editing/text_editing.dart

```diff
@@ -439,6 +440,142 @@ class AutofillInfo {
   }
 }
 
+/// Replaces a range of text in the original string with the text given in the
+/// replacement string.
+String _replace(String originalText, String replacementText, int start, int end) {
+  final String textStart = originalText.substring(0, start);
+  final String textEnd = originalText.substring(end, originalText.length);
+  final String newText = textStart + replacementText + textEnd;
+  return newText;
+}
+
+class TextEditingDeltaState {
+  const TextEditingDeltaState({
+    this.oldText = '',
+    this.deltaText = '',
+    this.deltaStart = -1,
+    this.deltaEnd = -1,
+    this.baseOffset = -1,
+    this.extentOffset = -1,
+    this.composingOffset = -1,
+    this.composingExtent = -1,
+  });
+
+  static TextEditingDeltaState inferDeltaState(EditingState newEditingState, EditingState? lastEditingState, TextEditingDeltaState? lastTextEditingDeltaState) {
```

Comment:
Would it make this any simpler if you used [inputType](https://developer.mozilla.org/en-US/docs/Web/API/InputEvent/inputType)?  Or is that not possible...

Reply:
I attempted to use `inputType` however there is some normalization I do in order to get the behavior of the Web to fit into the processing we do on the framework side that doesn't allow for this. 

For example, when trying to use `insertCompositionText` I discovered it is called for the entire composed word.
So say we have `hello|` and we delete character by character starting at the end. In this case `insertCompositionText` is called for every deletion. However on Android & iOS, when we delete the final character `h` this is reported as a regular deletion, and not one from the composing region.